// @vidstack/player/dist/dev/chunks/chunk.EZDBIN3S.js
import {
  __decorateClass
} from "./chunk.LNH2V2XS.js";

// src/ui/toggle-button/ToggleButtonElement.ts
import {
  eventListener,
  focusVisiblePolyfill,
  isKeyboardClick,
  isKeyboardEvent,
  logElementLifecycle,
  setAttribute,
  setAttributeIfEmpty
} from "@vidstack/foundation";
import {
  html,
  LitElement
} from "lit";
import { property } from "lit/decorators.js";

// src/ui/toggle-button/styles.ts
import { css } from "lit";
var toggleButtonElementStyles = css`
  :host {
    display: table;
    contain: content;
    user-select: none;
    cursor: pointer;
    -webkit-user-select: none;
    -webkit-tap-highlight-color: transparent;
  }

  :host([hidden]) {
    display: none;
  }

  :host(:focus) {
    outline: none;
  }
  :host(:focus-visible) {
    outline: 1px auto Highlight;
    outline: 1px auto -webkit-focus-ring-color;
  }
  :host(.focus-visible) {
    outline: 1px auto Highlight;
    outline: 1px auto -webkit-focus-ring-color;
  }
`;

// src/ui/toggle-button/ToggleButtonElement.ts
var ToggleButtonElement = class extends LitElement {
  constructor() {
    super();
    this.pressed = false;
    this.disabled = false;
    this._handleButtonClickCapture = eventListener(
      this,
      "click",
      (event) => {
        if (this.disabled) {
          event.preventDefault();
          event.stopImmediatePropagation();
        }
      },
      { capture: true }
    );
    if (true)
      logElementLifecycle(this);
    focusVisiblePolyfill(this);
    ["pointerup", "keydown"].forEach((eventType) => {
      eventListener(this, eventType, (event) => {
        if (this.disabled || isKeyboardEvent(event) && !isKeyboardClick(event)) {
          return;
        }
        this._handleButtonClick(event);
      });
    });
  }
  static get styles() {
    return [toggleButtonElementStyles];
  }
  static get parts() {
    return [];
  }
  connectedCallback() {
    super.connectedCallback();
    setAttributeIfEmpty(this, "tabindex", "0");
    setAttributeIfEmpty(this, "role", "button");
    this._updateAriaPressedAttr();
  }
  updated(changedProperties) {
    if (changedProperties.has("pressed")) {
      this._updateAriaPressedAttr();
    }
  }
  _updateAriaPressedAttr() {
    setAttribute(this, "aria-pressed", this.pressed ? "true" : "false");
  }
  render() {
    return this._renderDefaultSlot();
  }
  _renderDefaultSlot() {
    return html`<slot></slot>`;
  }
  _handleButtonClick(event) {
    this.pressed = !this.pressed;
  }
};
__decorateClass([
  property({ type: Boolean })
], ToggleButtonElement.prototype, "pressed", 2);
__decorateClass([
  property({ type: Boolean, reflect: true })
], ToggleButtonElement.prototype, "disabled", 2);

export {
  ToggleButtonElement
};
