// @vidstack/player/dist/node/index.js
import "@vidstack/foundation/shims/install-safe.js";

import {
  MediaSyncElement
} from "./chunks/chunk.6FEQRKGL.js";
import {
  MediaVisibilityElement
} from "./chunks/chunk.6TZ45PYH.js";
import {
  MediaController,
  MediaElement,
  UserIdleController,
  mediaDiscoveryId
} from "./chunks/chunk.VQVR7GTT.js";
import {
  TimeSliderElement
} from "./chunks/chunk.JJNEIAHO.js";
import {
  TimeElement
} from "./chunks/chunk.THUCYWQY.js";
import {
  VolumeSliderElement
} from "./chunks/chunk.RJA3LGSL.js";
import {
  SliderElement,
  SliderKeyDirection
} from "./chunks/chunk.T2GQIT5D.js";
import {
  MuteButtonElement
} from "./chunks/chunk.FCD2TYCA.js";
import {
  PlayButtonElement
} from "./chunks/chunk.XPLBBOMG.js";
import {
  PosterElement
} from "./chunks/chunk.NTPT5EBM.js";
import {
  SliderValueTextElement
} from "./chunks/chunk.ZCV5RCAM.js";
import {
  SliderVideoElement
} from "./chunks/chunk.4GKVTJUB.js";
import {
  createSliderStore,
  sliderStoreContext
} from "./chunks/chunk.CQHSYXO3.js";
import {
  AspectRatioElement
} from "./chunks/chunk.MD6WMDT5.js";
import "./chunks/chunk.LICA45E5.js";
import {
  AudioElement
} from "./chunks/chunk.DOZ4IU4N.js";
import {
  FullscreenButtonElement
} from "./chunks/chunk.QMLLKCOK.js";
import {
  ToggleButtonElement
} from "./chunks/chunk.IG3G5SFJ.js";
import {
  MediaRemoteControl
} from "./chunks/chunk.VHSBPZK3.js";
import {
  GestureElement
} from "./chunks/chunk.Z2KRCUL4.js";
import {
  mediaProviderElementContext
} from "./chunks/chunk.APMKZUJJ.js";
import {
  HLS_EXTENSIONS,
  HLS_TYPES,
  HlsElement
} from "./chunks/chunk.CP3EQARV.js";
import {
  VideoElement,
  VideoFullscreenController,
  VideoPresentationController
} from "./chunks/chunk.ANTAZFQD.js";
import {
  AUDIO_EXTENSIONS,
  Html5MediaElement,
  VIDEO_EXTENSIONS
} from "./chunks/chunk.F3A3LGBW.js";
import {
  MediaProviderElement,
  mediaProviderDiscoveryId
} from "./chunks/chunk.V5MIBM3O.js";
import {
  MEDIA_STORE_DEFAULTS,
  MediaType,
  ViewType,
  createMediaStore,
  mediaStoreContext,
  mediaStoreSubscription,
  resetMediaStore,
  softResetMediaStore
} from "./chunks/chunk.O2H47C42.js";
import "./chunks/chunk.NUDNXB7T.js";

// src/media/interact/media-event-listener.ts
import { discover, listen } from "@vidstack/foundation";
function mediaEventListener(host, eventType, listener) {
  discover(host, mediaProviderDiscoveryId, (provider, onDisconnect) => {
    const off = listen(provider, eventType, listener);
    onDisconnect(() => {
      off == null ? void 0 : off();
    });
  });
}

// src/media/MediaError.ts
var MediaErrorCode = /* @__PURE__ */ ((MediaErrorCode2) => {
  MediaErrorCode2[MediaErrorCode2["Aborted"] = 1] = "Aborted";
  MediaErrorCode2[MediaErrorCode2["Network"] = 2] = "Network";
  MediaErrorCode2[MediaErrorCode2["Decode"] = 3] = "Decode";
  MediaErrorCode2[MediaErrorCode2["SrcNotSupported"] = 4] = "SrcNotSupported";
  return MediaErrorCode2;
})(MediaErrorCode || {});

// src/index.ts
import {
  appendTriggerEvent,
  findTriggerEvent,
  hasTriggerEvent,
  isHlsjsSupported,
  walkTriggerEventChain
} from "@vidstack/foundation";
if (false) {
  console.warn("`@vidstack/player` is in dev mode. Not recommended for production!");
}
export {
  AUDIO_EXTENSIONS,
  AspectRatioElement,
  AudioElement,
  FullscreenButtonElement,
  GestureElement,
  HLS_EXTENSIONS,
  HLS_TYPES,
  HlsElement,
  Html5MediaElement,
  MEDIA_STORE_DEFAULTS,
  MediaController,
  MediaElement,
  MediaErrorCode,
  MediaProviderElement,
  MediaRemoteControl,
  MediaSyncElement,
  MediaType,
  MediaVisibilityElement,
  MuteButtonElement,
  PlayButtonElement,
  PosterElement,
  SliderElement,
  SliderKeyDirection,
  SliderValueTextElement,
  SliderVideoElement,
  TimeElement,
  TimeSliderElement,
  ToggleButtonElement,
  UserIdleController,
  VIDEO_EXTENSIONS,
  VideoElement,
  VideoFullscreenController,
  VideoPresentationController,
  ViewType,
  VolumeSliderElement,
  appendTriggerEvent,
  createMediaStore,
  createSliderStore,
  findTriggerEvent,
  hasTriggerEvent,
  isHlsjsSupported,
  mediaDiscoveryId,
  mediaEventListener,
  mediaProviderDiscoveryId,
  mediaProviderElementContext,
  mediaStoreContext,
  mediaStoreSubscription,
  resetMediaStore,
  sliderStoreContext,
  softResetMediaStore,
  walkTriggerEventChain
};
