// @vidstack/player/dist/dev/index.js
import {
  MediaSyncElement
} from "./chunks/chunk.EBAL4TQS.js";
import {
  MediaVisibilityElement
} from "./chunks/chunk.ZXXE6NEW.js";
import {
  MediaController,
  MediaElement,
  UserIdleController,
  mediaDiscoveryId
} from "./chunks/chunk.KT56YXMN.js";
import {
  TimeSliderElement
} from "./chunks/chunk.EMEGDAP5.js";
import {
  TimeElement
} from "./chunks/chunk.AKP4QMGQ.js";
import {
  VolumeSliderElement
} from "./chunks/chunk.6KQJYMWG.js";
import {
  SliderElement,
  SliderKeyDirection
} from "./chunks/chunk.ANNPV6ES.js";
import {
  MuteButtonElement
} from "./chunks/chunk.BPOEGC36.js";
import {
  PlayButtonElement
} from "./chunks/chunk.EFKVZB3S.js";
import {
  PosterElement
} from "./chunks/chunk.MFS4ZDZH.js";
import {
  SliderValueTextElement
} from "./chunks/chunk.URHGC37V.js";
import {
  SliderVideoElement
} from "./chunks/chunk.GLQMOABE.js";
import {
  createSliderStore,
  sliderStoreContext
} from "./chunks/chunk.CQHSYXO3.js";
import {
  AspectRatioElement
} from "./chunks/chunk.75662UYQ.js";
import {
  AudioElement
} from "./chunks/chunk.RL3MKEP7.js";
import {
  FullscreenButtonElement
} from "./chunks/chunk.TACPBOYA.js";
import {
  ToggleButtonElement
} from "./chunks/chunk.EZDBIN3S.js";
import {
  MediaRemoteControl
} from "./chunks/chunk.X5QPRJP2.js";
import {
  GestureElement
} from "./chunks/chunk.ZIGXS4GT.js";
import {
  mediaProviderElementContext
} from "./chunks/chunk.APMKZUJJ.js";
import {
  HLS_EXTENSIONS,
  HLS_TYPES,
  HlsElement
} from "./chunks/chunk.FSTKN4XK.js";
import {
  VideoElement,
  VideoFullscreenController,
  VideoPresentationController
} from "./chunks/chunk.WWUUJDIN.js";
import {
  AUDIO_EXTENSIONS,
  Html5MediaElement,
  VIDEO_EXTENSIONS
} from "./chunks/chunk.RVJZRTLD.js";
import {
  MediaProviderElement,
  mediaProviderDiscoveryId
} from "./chunks/chunk.BJN6YOPG.js";
import {
  MEDIA_STORE_DEFAULTS,
  MediaType,
  ViewType,
  createMediaStore,
  mediaStoreContext,
  mediaStoreSubscription,
  resetMediaStore,
  softResetMediaStore
} from "./chunks/chunk.O2H47C42.js";
import "./chunks/chunk.LNH2V2XS.js";

// src/media/interact/media-event-listener.ts
import { discover, listen } from "@vidstack/foundation";
function mediaEventListener(host, eventType, listener) {
  discover(host, mediaProviderDiscoveryId, (provider, onDisconnect) => {
    const off = listen(provider, eventType, listener);
    onDisconnect(() => {
      off?.();
    });
  });
}

// src/media/MediaError.ts
var MediaErrorCode = /* @__PURE__ */ ((MediaErrorCode2) => {
  MediaErrorCode2[MediaErrorCode2["Aborted"] = 1] = "Aborted";
  MediaErrorCode2[MediaErrorCode2["Network"] = 2] = "Network";
  MediaErrorCode2[MediaErrorCode2["Decode"] = 3] = "Decode";
  MediaErrorCode2[MediaErrorCode2["SrcNotSupported"] = 4] = "SrcNotSupported";
  return MediaErrorCode2;
})(MediaErrorCode || {});

// src/index.ts
import {
  appendTriggerEvent,
  findTriggerEvent,
  hasTriggerEvent,
  isHlsjsSupported,
  walkTriggerEventChain
} from "@vidstack/foundation";
if (true) {
  console.warn("`@vidstack/player` is in dev mode. Not recommended for production!");
}
export {
  AUDIO_EXTENSIONS,
  AspectRatioElement,
  AudioElement,
  FullscreenButtonElement,
  GestureElement,
  HLS_EXTENSIONS,
  HLS_TYPES,
  HlsElement,
  Html5MediaElement,
  MEDIA_STORE_DEFAULTS,
  MediaController,
  MediaElement,
  MediaErrorCode,
  MediaProviderElement,
  MediaRemoteControl,
  MediaSyncElement,
  MediaType,
  MediaVisibilityElement,
  MuteButtonElement,
  PlayButtonElement,
  PosterElement,
  SliderElement,
  SliderKeyDirection,
  SliderValueTextElement,
  SliderVideoElement,
  TimeElement,
  TimeSliderElement,
  ToggleButtonElement,
  UserIdleController,
  VIDEO_EXTENSIONS,
  VideoElement,
  VideoFullscreenController,
  VideoPresentationController,
  ViewType,
  VolumeSliderElement,
  appendTriggerEvent,
  createMediaStore,
  createSliderStore,
  findTriggerEvent,
  hasTriggerEvent,
  isHlsjsSupported,
  mediaDiscoveryId,
  mediaEventListener,
  mediaProviderDiscoveryId,
  mediaProviderElementContext,
  mediaStoreContext,
  mediaStoreSubscription,
  resetMediaStore,
  sliderStoreContext,
  softResetMediaStore,
  walkTriggerEventChain
};
