// @vidstack/player/dist/node/chunks/chunk.NTPT5EBM.js
import "@vidstack/foundation/shims/install-safe.js";

import {
  MediaRemoteControl
} from "./chunk.VHSBPZK3.js";
import {
  mediaStoreSubscription
} from "./chunk.O2H47C42.js";
import {
  __decorateClass,
  b,
  e,
  i,
  s,
  t,
  y
} from "./chunk.NUDNXB7T.js";

// src/ui/poster/PosterElement.ts
import { preconnect } from "@vidstack/foundation";

// ../../node_modules/.pnpm/lit-html@2.3.1/node_modules/lit-html/node/directives/if-defined.js
var l = (l2) => null != l2 ? l2 : b;

// src/ui/poster/PosterElement.ts
var didPreconnect = /* @__PURE__ */ new Set();
var PosterElement = class extends s {
  constructor() {
    super();
    this.__canLoad = false;
    this._mediaRemoteControl = new MediaRemoteControl(this);
    mediaStoreSubscription(this, "poster", ($poster) => {
      (typeof window !== "undefined" ? window : null)?.requestAnimationFrame(() => {
        if (!this.__canLoad && !didPreconnect.has($poster)) {
          preconnect($poster, "prefetch");
          didPreconnect.add($poster);
        }
      });
      this.__src = $poster;
    });
    mediaStoreSubscription(this, "canLoad", ($canLoad) => {
      this._handleCanLoadChange($canLoad);
      this.__canLoad = $canLoad;
    });
  }
  static get styles() {
    return [
      i`:host{display:block;contain:content;pointer-events:none;object-fit:cover;box-sizing:border-box}:host([hidden]){display:none}img{display:block;width:100%;height:100%;pointer-events:none;object-fit:inherit;object-position:inherit;user-select:none;-webkit-user-select:none;box-sizing:border-box}`
    ];
  }
  get src() {
    return this.__src;
  }
  connectedCallback() {
    super.connectedCallback();
    this._mediaRemoteControl.hidePoster();
  }
  disconnectedCallback() {
    this._mediaRemoteControl.showPoster();
    super.disconnectedCallback();
    this._handleCanLoadChange(false);
  }
  render() {
    return y`<img part="img" src="${l(this.__canLoad ? this.src : null)}" alt="${l(this.alt)}" @load="${this._handleImgLoad}" @error="${this._handleImgError}">`;
  }
  _setImgLoadingAttr() {
    this.removeAttribute("img-error");
    this.removeAttribute("img-loaded");
    if (this.src && this.src.length > 0) {
      this.setAttribute("img-loading", "");
    }
  }
  _handleImgLoad() {
    this.removeAttribute("img-loading");
    this.setAttribute("img-loaded", "");
  }
  _handleImgError() {
    this.removeAttribute("img-loading");
    this.setAttribute("img-error", "");
  }
  _handleCanLoadChange(canLoad) {
    if (canLoad) {
      this._setImgLoadingAttr();
    } else {
      this.removeAttribute("img-error");
      this.removeAttribute("img-loaded");
      this.removeAttribute("img-loading");
    }
  }
};
__decorateClass([
  t()
], PosterElement.prototype, "__src", 2);
__decorateClass([
  t()
], PosterElement.prototype, "__canLoad", 2);
__decorateClass([
  e()
], PosterElement.prototype, "alt", 2);

export {
  PosterElement
};
