// @vidstack/player/dist/node/chunks/chunk.4GKVTJUB.js
import "@vidstack/foundation/shims/install-safe.js";

import {
  sliderStoreContext
} from "./chunk.CQHSYXO3.js";
import {
  e as e2,
  i as i2,
  t as t2
} from "./chunk.LICA45E5.js";
import {
  __decorateClass,
  b,
  e,
  i,
  s,
  t,
  y,
  z
} from "./chunk.NUDNXB7T.js";

// src/ui/slider/SliderVideoElement.ts
import { ifNonEmpty, redispatchEvent, storeRecordSubscription } from "@vidstack/foundation";

// ../../node_modules/.pnpm/lit-html@2.3.1/node_modules/lit-html/node/directive-helpers.js
var { H: l } = z;
var e3 = (o3) => void 0 === o3.strings;

// ../../node_modules/.pnpm/lit-html@2.3.1/node_modules/lit-html/node/async-directive.js
var s2 = (i3, t3) => {
  var e5, o3;
  const r2 = i3._$AN;
  if (void 0 === r2)
    return false;
  for (const i4 of r2)
    null === (o3 = (e5 = i4)._$AO) || void 0 === o3 || o3.call(e5, t3, false), s2(i4, t3);
  return true;
};
var o = (i3) => {
  let t3, e5;
  do {
    if (void 0 === (t3 = i3._$AM))
      break;
    e5 = t3._$AN, e5.delete(i3), i3 = t3;
  } while (0 === (null == e5 ? void 0 : e5.size));
};
var r = (i3) => {
  for (let t3; t3 = i3._$AM; i3 = t3) {
    let e5 = t3._$AN;
    if (void 0 === e5)
      t3._$AN = e5 = /* @__PURE__ */ new Set();
    else if (e5.has(i3))
      break;
    e5.add(i3), l2(t3);
  }
};
function n(i3) {
  void 0 !== this._$AN ? (o(this), this._$AM = i3, r(this)) : this._$AM = i3;
}
function h(i3, t3 = false, e5 = 0) {
  const r2 = this._$AH, n3 = this._$AN;
  if (void 0 !== n3 && 0 !== n3.size)
    if (t3)
      if (Array.isArray(r2))
        for (let i4 = e5; i4 < r2.length; i4++)
          s2(r2[i4], false), o(r2[i4]);
      else
        null != r2 && (s2(r2, false), o(r2));
    else
      s2(this, i3);
}
var l2 = (i3) => {
  var t3, s3, o3, r2;
  i3.type == t2.CHILD && (null !== (t3 = (o3 = i3)._$AP) && void 0 !== t3 || (o3._$AP = h), null !== (s3 = (r2 = i3)._$AQ) && void 0 !== s3 || (r2._$AQ = n));
};
var c = class extends i2 {
  constructor() {
    super(...arguments), this._$AN = void 0;
  }
  _$AT(i3, t3, e5) {
    super._$AT(i3, t3, e5), r(this), this.isConnected = i3._$AU;
  }
  _$AO(i3, t3 = true) {
    var e5, r2;
    i3 !== this.isConnected && (this.isConnected = i3, i3 ? null === (e5 = this.reconnected) || void 0 === e5 || e5.call(this) : null === (r2 = this.disconnected) || void 0 === r2 || r2.call(this)), t3 && (s2(this, i3), o(this));
  }
  setValue(t3) {
    if (e3(this._$Ct))
      this._$Ct._$AI(t3, this);
    else {
      const i3 = [...this._$Ct._$AH];
      i3[this._$Ci] = t3, this._$Ct._$AI(i3, this, 0);
    }
  }
  disconnected() {
  }
  reconnected() {
  }
};

// ../../node_modules/.pnpm/lit-html@2.3.1/node_modules/lit-html/node/directives/ref.js
var e4 = () => new o2();
var o2 = class {
};
var h2 = /* @__PURE__ */ new WeakMap();
var n2 = e2(class extends c {
  render(t3) {
    return b;
  }
  update(t3, [s3]) {
    var e5;
    const o3 = s3 !== this.Y;
    return o3 && void 0 !== this.Y && this.rt(void 0), (o3 || this.lt !== this.dt) && (this.Y = s3, this.ct = null === (e5 = t3.options) || void 0 === e5 ? void 0 : e5.host, this.rt(this.dt = t3.element)), b;
  }
  rt(i3) {
    var t3;
    if ("function" == typeof this.Y) {
      const s3 = null !== (t3 = this.ct) && void 0 !== t3 ? t3 : globalThis;
      let e5 = h2.get(s3);
      void 0 === e5 && (e5 = /* @__PURE__ */ new WeakMap(), h2.set(s3, e5)), void 0 !== e5.get(this.Y) && this.Y.call(this.ct, void 0), e5.set(this.Y, i3), void 0 !== i3 && this.Y.call(this.ct, i3);
    } else
      this.Y.value = i3;
  }
  get lt() {
    var i3, t3, s3;
    return "function" == typeof this.Y ? null === (t3 = h2.get(null !== (i3 = this.ct) && void 0 !== i3 ? i3 : globalThis)) || void 0 === t3 ? void 0 : t3.get(this.Y) : null === (s3 = this.Y) || void 0 === s3 ? void 0 : s3.value;
  }
  disconnected() {
    this.lt === this.dt && this.rt(void 0);
  }
  reconnected() {
    this.rt(this.dt);
  }
});

// src/ui/slider/SliderVideoElement.ts
var SliderVideoElement = class extends s {
  constructor() {
    super();
    this._videoRef = e4();
    this.__canPlay = false;
    this.__hasError = false;
    storeRecordSubscription(this, sliderStoreContext, "pointerValue", ($previewTime) => {
      this._updateCurrentTime($previewTime);
    });
  }
  static get styles() {
    return [
      i`:host{display:inline-block;contain:content}:host([hidden]){display:none}video{display:block;width:100%;height:auto}`
    ];
  }
  get videoElement() {
    return this._videoRef.value;
  }
  _updateCurrentTime(seconds) {
    if (!this.__hasError && this.__canPlay && this.videoElement.currentTime !== seconds) {
      this.videoElement.currentTime = seconds;
    }
  }
  willUpdate(changedProperties) {
    if (changedProperties.has("src")) {
      this.__canPlay = false;
      this.__hasError = false;
      this.removeAttribute("video-can-play");
      this.removeAttribute("video-error");
    }
    super.willUpdate(changedProperties);
  }
  render() {
    return this._renderVideo();
  }
  _renderVideo() {
    return y`<video part="video" muted playsinline preload="auto" src="${ifNonEmpty(this.src)}" @canplay="${this._handleCanPlay}" @error="${this._handleError}" ${n2(this._videoRef)}></video>`;
  }
  async _handleCanPlay(event) {
    this.__canPlay = true;
    this.setAttribute("video-can-play", "");
    redispatchEvent(this, event);
  }
  _handleError(event) {
    this.__hasError = true;
    this.setAttribute("video-error", "");
    redispatchEvent(this, event);
  }
};
__decorateClass([
  e()
], SliderVideoElement.prototype, "src", 2);
__decorateClass([
  t()
], SliderVideoElement.prototype, "__canPlay", 2);
__decorateClass([
  t()
], SliderVideoElement.prototype, "__hasError", 2);

export {
  SliderVideoElement
};
