// @vidstack/player/dist/node/chunks/chunk.NUDNXB7T.js
import "@vidstack/foundation/shims/install-safe.js";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i4 = decorators.length - 1, decorator; i4 >= 0; i4--)
    if (decorator = decorators[i4])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result)
    __defProp(target, key, result);
  return result;
};

// ../../node_modules/.pnpm/@lit+reactive-element@1.4.1/node_modules/@lit/reactive-element/node/css-tag.js
var t = globalThis;
var e = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype;
var s = Symbol();
var n = /* @__PURE__ */ new WeakMap();
var o = class {
  constructor(t4, e6, n6) {
    if (this._$cssResult$ = true, n6 !== s)
      throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t4, this.t = e6;
  }
  get styleSheet() {
    let t4 = this.o;
    const s5 = this.t;
    if (e && void 0 === t4) {
      const e6 = void 0 !== s5 && 1 === s5.length;
      e6 && (t4 = n.get(s5)), void 0 === t4 && ((this.o = t4 = new CSSStyleSheet()).replaceSync(this.cssText), e6 && n.set(s5, t4));
    }
    return t4;
  }
  toString() {
    return this.cssText;
  }
};
var r = (t4) => new o("string" == typeof t4 ? t4 : t4 + "", void 0, s);
var i = (t4, ...e6) => {
  const n6 = 1 === t4.length ? t4[0] : e6.reduce((e7, s5, n7) => e7 + ((t5) => {
    if (true === t5._$cssResult$)
      return t5.cssText;
    if ("number" == typeof t5)
      return t5;
    throw Error("Value passed to 'css' function must be a 'css' function result: " + t5 + ". Use 'unsafeCSS' to pass non-literal values, but take care to ensure page security.");
  })(s5) + t4[n7 + 1], t4[0]);
  return new o(n6, t4, s);
};
var S = (s5, n6) => {
  e ? s5.adoptedStyleSheets = n6.map((t4) => t4 instanceof CSSStyleSheet ? t4 : t4.styleSheet) : n6.forEach((e6) => {
    const n7 = document.createElement("style"), o6 = t.litNonce;
    void 0 !== o6 && n7.setAttribute("nonce", o6), n7.textContent = e6.cssText, s5.appendChild(n7);
  });
};
var c = e || void 0 === t.CSSStyleSheet ? (t4) => t4 : (t4) => t4 instanceof CSSStyleSheet ? ((t5) => {
  let e6 = "";
  for (const s5 of t5.cssRules)
    e6 += s5.cssText;
  return r(e6);
})(t4) : t4;

// ../../node_modules/.pnpm/@lit+reactive-element@1.4.1/node_modules/@lit/reactive-element/node/reactive-element.js
var s2;
var e2;
var r2 = globalThis;
null !== (s2 = r2.customElements) && void 0 !== s2 || (r2.customElements = { define() {
} });
var h = r2.trustedTypes;
var o2 = h ? h.emptyScript : "";
var n2 = r2.reactiveElementPolyfillSupport;
var l = { toAttribute(t4, i4) {
  switch (i4) {
    case Boolean:
      t4 = t4 ? o2 : null;
      break;
    case Object:
    case Array:
      t4 = null == t4 ? t4 : JSON.stringify(t4);
  }
  return t4;
}, fromAttribute(t4, i4) {
  let s5 = t4;
  switch (i4) {
    case Boolean:
      s5 = null !== t4;
      break;
    case Number:
      s5 = null === t4 ? null : Number(t4);
      break;
    case Object:
    case Array:
      try {
        s5 = JSON.parse(t4);
      } catch (t5) {
        s5 = null;
      }
  }
  return s5;
} };
var a = (t4, i4) => i4 !== t4 && (i4 == i4 || t4 == t4);
var d = { attribute: true, type: String, converter: l, reflect: false, hasChanged: a };
var u = void 0 === r2.HTMLElement;
u && (r2.HTMLElement = class {
});
var c2 = class extends HTMLElement {
  constructor() {
    super(), this._$Ei = /* @__PURE__ */ new Map(), this.isUpdatePending = false, this.hasUpdated = false, this._$El = null, this.u();
  }
  static addInitializer(t4) {
    var i4;
    null !== (i4 = this.h) && void 0 !== i4 || (this.h = []), this.h.push(t4);
  }
  static get observedAttributes() {
    this.finalize();
    const t4 = [];
    return this.elementProperties.forEach((i4, s5) => {
      const e6 = this._$Ep(s5, i4);
      void 0 !== e6 && (this._$Ev.set(e6, s5), t4.push(e6));
    }), t4;
  }
  static createProperty(t4, i4 = d) {
    if (i4.state && (i4.attribute = false), this.finalize(), this.elementProperties.set(t4, i4), !i4.noAccessor && !this.prototype.hasOwnProperty(t4)) {
      const s5 = "symbol" == typeof t4 ? Symbol() : "__" + t4, e6 = this.getPropertyDescriptor(t4, s5, i4);
      void 0 !== e6 && Object.defineProperty(this.prototype, t4, e6);
    }
  }
  static getPropertyDescriptor(t4, i4, s5) {
    return { get() {
      return this[i4];
    }, set(e6) {
      const r4 = this[t4];
      this[i4] = e6, this.requestUpdate(t4, r4, s5);
    }, configurable: true, enumerable: true };
  }
  static getPropertyOptions(t4) {
    return this.elementProperties.get(t4) || d;
  }
  static finalize() {
    if (this.hasOwnProperty("finalized"))
      return false;
    this.finalized = true;
    const t4 = Object.getPrototypeOf(this);
    if (t4.finalize(), this.elementProperties = new Map(t4.elementProperties), this._$Ev = /* @__PURE__ */ new Map(), this.hasOwnProperty("properties")) {
      const t5 = this.properties, i4 = [...Object.getOwnPropertyNames(t5), ...Object.getOwnPropertySymbols(t5)];
      for (const s5 of i4)
        this.createProperty(s5, t5[s5]);
    }
    return this.elementStyles = this.finalizeStyles(this.styles), true;
  }
  static finalizeStyles(i4) {
    const s5 = [];
    if (Array.isArray(i4)) {
      const e6 = new Set(i4.flat(1 / 0).reverse());
      for (const i5 of e6)
        s5.unshift(c(i5));
    } else
      void 0 !== i4 && s5.push(c(i4));
    return s5;
  }
  static _$Ep(t4, i4) {
    const s5 = i4.attribute;
    return false === s5 ? void 0 : "string" == typeof s5 ? s5 : "string" == typeof t4 ? t4.toLowerCase() : void 0;
  }
  u() {
    var t4;
    this._$E_ = new Promise((t5) => this.enableUpdating = t5), this._$AL = /* @__PURE__ */ new Map(), this._$Eg(), this.requestUpdate(), null === (t4 = this.constructor.h) || void 0 === t4 || t4.forEach((t5) => t5(this));
  }
  addController(t4) {
    var i4, s5;
    (null !== (i4 = this._$ES) && void 0 !== i4 ? i4 : this._$ES = []).push(t4), void 0 !== this.renderRoot && this.isConnected && (null === (s5 = t4.hostConnected) || void 0 === s5 || s5.call(t4));
  }
  removeController(t4) {
    var i4;
    null === (i4 = this._$ES) || void 0 === i4 || i4.splice(this._$ES.indexOf(t4) >>> 0, 1);
  }
  _$Eg() {
    this.constructor.elementProperties.forEach((t4, i4) => {
      this.hasOwnProperty(i4) && (this._$Ei.set(i4, this[i4]), delete this[i4]);
    });
  }
  createRenderRoot() {
    var t4;
    const s5 = null !== (t4 = this.shadowRoot) && void 0 !== t4 ? t4 : this.attachShadow(this.constructor.shadowRootOptions);
    return S(s5, this.constructor.elementStyles), s5;
  }
  connectedCallback() {
    var t4;
    void 0 === this.renderRoot && (this.renderRoot = this.createRenderRoot()), this.enableUpdating(true), null === (t4 = this._$ES) || void 0 === t4 || t4.forEach((t5) => {
      var i4;
      return null === (i4 = t5.hostConnected) || void 0 === i4 ? void 0 : i4.call(t5);
    });
  }
  enableUpdating(t4) {
  }
  disconnectedCallback() {
    var t4;
    null === (t4 = this._$ES) || void 0 === t4 || t4.forEach((t5) => {
      var i4;
      return null === (i4 = t5.hostDisconnected) || void 0 === i4 ? void 0 : i4.call(t5);
    });
  }
  attributeChangedCallback(t4, i4, s5) {
    this._$AK(t4, s5);
  }
  _$EO(t4, i4, s5 = d) {
    var e6;
    const r4 = this.constructor._$Ep(t4, s5);
    if (void 0 !== r4 && true === s5.reflect) {
      const h3 = (void 0 !== (null === (e6 = s5.converter) || void 0 === e6 ? void 0 : e6.toAttribute) ? s5.converter : l).toAttribute(i4, s5.type);
      this._$El = t4, null == h3 ? this.removeAttribute(r4) : this.setAttribute(r4, h3), this._$El = null;
    }
  }
  _$AK(t4, i4) {
    var s5;
    const e6 = this.constructor, r4 = e6._$Ev.get(t4);
    if (void 0 !== r4 && this._$El !== r4) {
      const t5 = e6.getPropertyOptions(r4), h3 = "function" == typeof t5.converter ? { fromAttribute: t5.converter } : void 0 !== (null === (s5 = t5.converter) || void 0 === s5 ? void 0 : s5.fromAttribute) ? t5.converter : l;
      this._$El = r4, this[r4] = h3.fromAttribute(i4, t5.type), this._$El = null;
    }
  }
  requestUpdate(t4, i4, s5) {
    let e6 = true;
    void 0 !== t4 && (((s5 = s5 || this.constructor.getPropertyOptions(t4)).hasChanged || a)(this[t4], i4) ? (this._$AL.has(t4) || this._$AL.set(t4, i4), true === s5.reflect && this._$El !== t4 && (void 0 === this._$EC && (this._$EC = /* @__PURE__ */ new Map()), this._$EC.set(t4, s5))) : e6 = false), !this.isUpdatePending && e6 && (this._$E_ = this._$Ej());
  }
  async _$Ej() {
    this.isUpdatePending = true;
    try {
      await this._$E_;
    } catch (t5) {
      Promise.reject(t5);
    }
    const t4 = this.scheduleUpdate();
    return null != t4 && await t4, !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    var t4;
    if (!this.isUpdatePending)
      return;
    this.hasUpdated, this._$Ei && (this._$Ei.forEach((t5, i5) => this[i5] = t5), this._$Ei = void 0);
    let i4 = false;
    const s5 = this._$AL;
    try {
      i4 = this.shouldUpdate(s5), i4 ? (this.willUpdate(s5), null === (t4 = this._$ES) || void 0 === t4 || t4.forEach((t5) => {
        var i5;
        return null === (i5 = t5.hostUpdate) || void 0 === i5 ? void 0 : i5.call(t5);
      }), this.update(s5)) : this._$Ek();
    } catch (t5) {
      throw i4 = false, this._$Ek(), t5;
    }
    i4 && this._$AE(s5);
  }
  willUpdate(t4) {
  }
  _$AE(t4) {
    var i4;
    null === (i4 = this._$ES) || void 0 === i4 || i4.forEach((t5) => {
      var i5;
      return null === (i5 = t5.hostUpdated) || void 0 === i5 ? void 0 : i5.call(t5);
    }), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t4)), this.updated(t4);
  }
  _$Ek() {
    this._$AL = /* @__PURE__ */ new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$E_;
  }
  shouldUpdate(t4) {
    return true;
  }
  update(t4) {
    void 0 !== this._$EC && (this._$EC.forEach((t5, i4) => this._$EO(i4, this[i4], t5)), this._$EC = void 0), this._$Ek();
  }
  updated(t4) {
  }
  firstUpdated(t4) {
  }
};
c2.finalized = true, c2.elementProperties = /* @__PURE__ */ new Map(), c2.elementStyles = [], c2.shadowRootOptions = { mode: "open" }, u && delete r2.HTMLElement, null == n2 || n2({ ReactiveElement: c2 }), (null !== (e2 = r2.reactiveElementVersions) && void 0 !== e2 ? e2 : r2.reactiveElementVersions = []).push("1.4.1");

// ../../node_modules/.pnpm/lit-html@2.3.1/node_modules/lit-html/node/lit-html.js
var t2;
var i2 = globalThis;
var s3 = i2.trustedTypes;
var e3 = s3 ? s3.createPolicy("lit-html", { createHTML: (t4) => t4 }) : void 0;
var o3 = `lit$${(Math.random() + "").slice(9)}$`;
var n3 = "?" + o3;
var l2 = `<${n3}>`;
var h2 = void 0 === i2.document ? { createTreeWalker: () => ({}) } : document;
var r3 = (t4 = "") => h2.createComment(t4);
var d2 = (t4) => null === t4 || "object" != typeof t4 && "function" != typeof t4;
var u2 = Array.isArray;
var c3 = (t4) => u2(t4) || "function" == typeof (null == t4 ? void 0 : t4[Symbol.iterator]);
var v = /<(?:(!--|\/[^a-zA-Z])|(\/?[a-zA-Z][^>\s]*)|(\/?$))/g;
var a2 = /-->/g;
var f = />/g;
var _ = RegExp(`>|[ 	
\f\r](?:([^\\s"'>=/]+)([ 	
\f\r]*=[ 	
\f\r]*(?:[^ 	
\f\r"'\`<>=]|("|')|))|$)`, "g");
var m = /'/g;
var p = /"/g;
var g = /^(?:script|style|textarea|title)$/i;
var $ = (t4) => (i4, ...s5) => ({ _$litType$: t4, strings: i4, values: s5 });
var y = $(1);
var x = $(2);
var T = Symbol.for("lit-noChange");
var b = Symbol.for("lit-nothing");
var w = /* @__PURE__ */ new WeakMap();
var A = (t4, i4, s5) => {
  var e6, o6;
  const n6 = null !== (e6 = null == s5 ? void 0 : s5.renderBefore) && void 0 !== e6 ? e6 : i4;
  let l5 = n6._$litPart$;
  if (void 0 === l5) {
    const t5 = null !== (o6 = null == s5 ? void 0 : s5.renderBefore) && void 0 !== o6 ? o6 : null;
    n6._$litPart$ = l5 = new S2(i4.insertBefore(r3(), t5), t5, void 0, null != s5 ? s5 : {});
  }
  return l5._$AI(t4), l5;
};
var E = h2.createTreeWalker(h2, 129, null, false);
var C = (t4, i4) => {
  const s5 = t4.length - 1, n6 = [];
  let h3, r4 = 2 === i4 ? "<svg>" : "", d3 = v;
  for (let i5 = 0; i5 < s5; i5++) {
    const s6 = t4[i5];
    let e6, u4, c4 = -1, $2 = 0;
    for (; $2 < s6.length && (d3.lastIndex = $2, u4 = d3.exec(s6), null !== u4); )
      $2 = d3.lastIndex, d3 === v ? "!--" === u4[1] ? d3 = a2 : void 0 !== u4[1] ? d3 = f : void 0 !== u4[2] ? (g.test(u4[2]) && (h3 = RegExp("</" + u4[2], "g")), d3 = _) : void 0 !== u4[3] && (d3 = _) : d3 === _ ? ">" === u4[0] ? (d3 = null != h3 ? h3 : v, c4 = -1) : void 0 === u4[1] ? c4 = -2 : (c4 = d3.lastIndex - u4[2].length, e6 = u4[1], d3 = void 0 === u4[3] ? _ : '"' === u4[3] ? p : m) : d3 === p || d3 === m ? d3 = _ : d3 === a2 || d3 === f ? d3 = v : (d3 = _, h3 = void 0);
    const y2 = d3 === _ && t4[i5 + 1].startsWith("/>") ? " " : "";
    r4 += d3 === v ? s6 + l2 : c4 >= 0 ? (n6.push(e6), s6.slice(0, c4) + "$lit$" + s6.slice(c4) + o3 + y2) : s6 + o3 + (-2 === c4 ? (n6.push(void 0), i5) : y2);
  }
  const u3 = r4 + (t4[s5] || "<?>") + (2 === i4 ? "</svg>" : "");
  if (!Array.isArray(t4) || !t4.hasOwnProperty("raw"))
    throw Error("invalid template strings array");
  return [void 0 !== e3 ? e3.createHTML(u3) : u3, n6];
};
var P = class {
  constructor({ strings: t4, _$litType$: i4 }, e6) {
    let l5;
    this.parts = [];
    let h3 = 0, d3 = 0;
    const u3 = t4.length - 1, c4 = this.parts, [v2, a3] = C(t4, i4);
    if (this.el = P.createElement(v2, e6), E.currentNode = this.el.content, 2 === i4) {
      const t5 = this.el.content, i5 = t5.firstChild;
      i5.remove(), t5.append(...i5.childNodes);
    }
    for (; null !== (l5 = E.nextNode()) && c4.length < u3; ) {
      if (1 === l5.nodeType) {
        if (l5.hasAttributes()) {
          const t5 = [];
          for (const i5 of l5.getAttributeNames())
            if (i5.endsWith("$lit$") || i5.startsWith(o3)) {
              const s5 = a3[d3++];
              if (t5.push(i5), void 0 !== s5) {
                const t6 = l5.getAttribute(s5.toLowerCase() + "$lit$").split(o3), i6 = /([.?@])?(.*)/.exec(s5);
                c4.push({ type: 1, index: h3, name: i6[2], strings: t6, ctor: "." === i6[1] ? k : "?" === i6[1] ? H : "@" === i6[1] ? I : M });
              } else
                c4.push({ type: 6, index: h3 });
            }
          for (const i5 of t5)
            l5.removeAttribute(i5);
        }
        if (g.test(l5.tagName)) {
          const t5 = l5.textContent.split(o3), i5 = t5.length - 1;
          if (i5 > 0) {
            l5.textContent = s3 ? s3.emptyScript : "";
            for (let s5 = 0; s5 < i5; s5++)
              l5.append(t5[s5], r3()), E.nextNode(), c4.push({ type: 2, index: ++h3 });
            l5.append(t5[i5], r3());
          }
        }
      } else if (8 === l5.nodeType)
        if (l5.data === n3)
          c4.push({ type: 2, index: h3 });
        else {
          let t5 = -1;
          for (; -1 !== (t5 = l5.data.indexOf(o3, t5 + 1)); )
            c4.push({ type: 7, index: h3 }), t5 += o3.length - 1;
        }
      h3++;
    }
  }
  static createElement(t4, i4) {
    const s5 = h2.createElement("template");
    return s5.innerHTML = t4, s5;
  }
};
function V(t4, i4, s5 = t4, e6) {
  var o6, n6, l5, h3;
  if (i4 === T)
    return i4;
  let r4 = void 0 !== e6 ? null === (o6 = s5._$Cl) || void 0 === o6 ? void 0 : o6[e6] : s5._$Cu;
  const u3 = d2(i4) ? void 0 : i4._$litDirective$;
  return (null == r4 ? void 0 : r4.constructor) !== u3 && (null === (n6 = null == r4 ? void 0 : r4._$AO) || void 0 === n6 || n6.call(r4, false), void 0 === u3 ? r4 = void 0 : (r4 = new u3(t4), r4._$AT(t4, s5, e6)), void 0 !== e6 ? (null !== (l5 = (h3 = s5)._$Cl) && void 0 !== l5 ? l5 : h3._$Cl = [])[e6] = r4 : s5._$Cu = r4), void 0 !== r4 && (i4 = V(t4, r4._$AS(t4, i4.values), r4, e6)), i4;
}
var N = class {
  constructor(t4, i4) {
    this.v = [], this._$AN = void 0, this._$AD = t4, this._$AM = i4;
  }
  get parentNode() {
    return this._$AM.parentNode;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  p(t4) {
    var i4;
    const { el: { content: s5 }, parts: e6 } = this._$AD, o6 = (null !== (i4 = null == t4 ? void 0 : t4.creationScope) && void 0 !== i4 ? i4 : h2).importNode(s5, true);
    E.currentNode = o6;
    let n6 = E.nextNode(), l5 = 0, r4 = 0, d3 = e6[0];
    for (; void 0 !== d3; ) {
      if (l5 === d3.index) {
        let i5;
        2 === d3.type ? i5 = new S2(n6, n6.nextSibling, this, t4) : 1 === d3.type ? i5 = new d3.ctor(n6, d3.name, d3.strings, this, t4) : 6 === d3.type && (i5 = new L(n6, this, t4)), this.v.push(i5), d3 = e6[++r4];
      }
      l5 !== (null == d3 ? void 0 : d3.index) && (n6 = E.nextNode(), l5++);
    }
    return o6;
  }
  m(t4) {
    let i4 = 0;
    for (const s5 of this.v)
      void 0 !== s5 && (void 0 !== s5.strings ? (s5._$AI(t4, s5, i4), i4 += s5.strings.length - 2) : s5._$AI(t4[i4])), i4++;
  }
};
var S2 = class {
  constructor(t4, i4, s5, e6) {
    var o6;
    this.type = 2, this._$AH = b, this._$AN = void 0, this._$AA = t4, this._$AB = i4, this._$AM = s5, this.options = e6, this._$C_ = null === (o6 = null == e6 ? void 0 : e6.isConnected) || void 0 === o6 || o6;
  }
  get _$AU() {
    var t4, i4;
    return null !== (i4 = null === (t4 = this._$AM) || void 0 === t4 ? void 0 : t4._$AU) && void 0 !== i4 ? i4 : this._$C_;
  }
  get parentNode() {
    let t4 = this._$AA.parentNode;
    const i4 = this._$AM;
    return void 0 !== i4 && 11 === t4.nodeType && (t4 = i4.parentNode), t4;
  }
  get startNode() {
    return this._$AA;
  }
  get endNode() {
    return this._$AB;
  }
  _$AI(t4, i4 = this) {
    t4 = V(this, t4, i4), d2(t4) ? t4 === b || null == t4 || "" === t4 ? (this._$AH !== b && this._$AR(), this._$AH = b) : t4 !== this._$AH && t4 !== T && this.$(t4) : void 0 !== t4._$litType$ ? this.T(t4) : void 0 !== t4.nodeType ? this.k(t4) : c3(t4) ? this.O(t4) : this.$(t4);
  }
  S(t4, i4 = this._$AB) {
    return this._$AA.parentNode.insertBefore(t4, i4);
  }
  k(t4) {
    this._$AH !== t4 && (this._$AR(), this._$AH = this.S(t4));
  }
  $(t4) {
    this._$AH !== b && d2(this._$AH) ? this._$AA.nextSibling.data = t4 : this.k(h2.createTextNode(t4)), this._$AH = t4;
  }
  T(t4) {
    var i4;
    const { values: s5, _$litType$: e6 } = t4, o6 = "number" == typeof e6 ? this._$AC(t4) : (void 0 === e6.el && (e6.el = P.createElement(e6.h, this.options)), e6);
    if ((null === (i4 = this._$AH) || void 0 === i4 ? void 0 : i4._$AD) === o6)
      this._$AH.m(s5);
    else {
      const t5 = new N(o6, this), i5 = t5.p(this.options);
      t5.m(s5), this.k(i5), this._$AH = t5;
    }
  }
  _$AC(t4) {
    let i4 = w.get(t4.strings);
    return void 0 === i4 && w.set(t4.strings, i4 = new P(t4)), i4;
  }
  O(t4) {
    u2(this._$AH) || (this._$AH = [], this._$AR());
    const i4 = this._$AH;
    let s5, e6 = 0;
    for (const o6 of t4)
      e6 === i4.length ? i4.push(s5 = new S2(this.S(r3()), this.S(r3()), this, this.options)) : s5 = i4[e6], s5._$AI(o6), e6++;
    e6 < i4.length && (this._$AR(s5 && s5._$AB.nextSibling, e6), i4.length = e6);
  }
  _$AR(t4 = this._$AA.nextSibling, i4) {
    var s5;
    for (null === (s5 = this._$AP) || void 0 === s5 || s5.call(this, false, true, i4); t4 && t4 !== this._$AB; ) {
      const i5 = t4.nextSibling;
      t4.remove(), t4 = i5;
    }
  }
  setConnected(t4) {
    var i4;
    void 0 === this._$AM && (this._$C_ = t4, null === (i4 = this._$AP) || void 0 === i4 || i4.call(this, t4));
  }
};
var M = class {
  constructor(t4, i4, s5, e6, o6) {
    this.type = 1, this._$AH = b, this._$AN = void 0, this.element = t4, this.name = i4, this._$AM = e6, this.options = o6, s5.length > 2 || "" !== s5[0] || "" !== s5[1] ? (this._$AH = Array(s5.length - 1).fill(new String()), this.strings = s5) : this._$AH = b;
  }
  get tagName() {
    return this.element.tagName;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AI(t4, i4 = this, s5, e6) {
    const o6 = this.strings;
    let n6 = false;
    if (void 0 === o6)
      t4 = V(this, t4, i4, 0), n6 = !d2(t4) || t4 !== this._$AH && t4 !== T, n6 && (this._$AH = t4);
    else {
      const e7 = t4;
      let l5, h3;
      for (t4 = o6[0], l5 = 0; l5 < o6.length - 1; l5++)
        h3 = V(this, e7[s5 + l5], i4, l5), h3 === T && (h3 = this._$AH[l5]), n6 || (n6 = !d2(h3) || h3 !== this._$AH[l5]), h3 === b ? t4 = b : t4 !== b && (t4 += (null != h3 ? h3 : "") + o6[l5 + 1]), this._$AH[l5] = h3;
    }
    n6 && !e6 && this.P(t4);
  }
  P(t4) {
    t4 === b ? this.element.removeAttribute(this.name) : this.element.setAttribute(this.name, null != t4 ? t4 : "");
  }
};
var k = class extends M {
  constructor() {
    super(...arguments), this.type = 3;
  }
  P(t4) {
    this.element[this.name] = t4 === b ? void 0 : t4;
  }
};
var R = s3 ? s3.emptyScript : "";
var H = class extends M {
  constructor() {
    super(...arguments), this.type = 4;
  }
  P(t4) {
    t4 && t4 !== b ? this.element.setAttribute(this.name, R) : this.element.removeAttribute(this.name);
  }
};
var I = class extends M {
  constructor(t4, i4, s5, e6, o6) {
    super(t4, i4, s5, e6, o6), this.type = 5;
  }
  _$AI(t4, i4 = this) {
    var s5;
    if ((t4 = null !== (s5 = V(this, t4, i4, 0)) && void 0 !== s5 ? s5 : b) === T)
      return;
    const e6 = this._$AH, o6 = t4 === b && e6 !== b || t4.capture !== e6.capture || t4.once !== e6.once || t4.passive !== e6.passive, n6 = t4 !== b && (e6 === b || o6);
    o6 && this.element.removeEventListener(this.name, this, e6), n6 && this.element.addEventListener(this.name, this, t4), this._$AH = t4;
  }
  handleEvent(t4) {
    var i4, s5;
    "function" == typeof this._$AH ? this._$AH.call(null !== (s5 = null === (i4 = this.options) || void 0 === i4 ? void 0 : i4.host) && void 0 !== s5 ? s5 : this.element, t4) : this._$AH.handleEvent(t4);
  }
};
var L = class {
  constructor(t4, i4, s5) {
    this.element = t4, this.type = 6, this._$AN = void 0, this._$AM = i4, this.options = s5;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AI(t4) {
    V(this, t4);
  }
};
var z = { A: "$lit$", M: o3, C: n3, L: 1, R: C, D: N, V: c3, I: V, H: S2, N: M, U: H, B: I, F: k, W: L };
var W = i2.litHtmlPolyfillSupport;
null == W || W(P, S2), (null !== (t2 = i2.litHtmlVersions) && void 0 !== t2 ? t2 : i2.litHtmlVersions = []).push("2.3.1");

// ../../node_modules/.pnpm/lit-element@3.2.2/node_modules/lit-element/lit-element.js
var l3;
var o4;
var s4 = class extends c2 {
  constructor() {
    super(...arguments), this.renderOptions = { host: this }, this._$Do = void 0;
  }
  createRenderRoot() {
    var t4, e6;
    const i4 = super.createRenderRoot();
    return null !== (t4 = (e6 = this.renderOptions).renderBefore) && void 0 !== t4 || (e6.renderBefore = i4.firstChild), i4;
  }
  update(t4) {
    const i4 = this.render();
    this.hasUpdated || (this.renderOptions.isConnected = this.isConnected), super.update(t4), this._$Do = A(i4, this.renderRoot, this.renderOptions);
  }
  connectedCallback() {
    var t4;
    super.connectedCallback(), null === (t4 = this._$Do) || void 0 === t4 || t4.setConnected(true);
  }
  disconnectedCallback() {
    var t4;
    super.disconnectedCallback(), null === (t4 = this._$Do) || void 0 === t4 || t4.setConnected(false);
  }
  render() {
    return T;
  }
};
s4.finalized = true, s4._$litElement$ = true, null === (l3 = globalThis.litElementHydrateSupport) || void 0 === l3 || l3.call(globalThis, { LitElement: s4 });
var n4 = globalThis.litElementPolyfillSupport;
null == n4 || n4({ LitElement: s4 });
(null !== (o4 = globalThis.litElementVersions) && void 0 !== o4 ? o4 : globalThis.litElementVersions = []).push("3.2.2");

// ../../node_modules/.pnpm/@lit+reactive-element@1.4.1/node_modules/@lit/reactive-element/node/decorators/property.js
var i3 = (i4, e6) => "method" === e6.kind && e6.descriptor && !("value" in e6.descriptor) ? { ...e6, finisher(n6) {
  n6.createProperty(e6.key, i4);
} } : { kind: "field", key: Symbol(), placement: "own", descriptor: {}, originalKey: e6.key, initializer() {
  "function" == typeof e6.initializer && (this[e6.key] = e6.initializer.call(this));
}, finisher(n6) {
  n6.createProperty(e6.key, i4);
} };
function e4(e6) {
  return (n6, t4) => void 0 !== t4 ? ((i4, e7, n7) => {
    e7.constructor.createProperty(n7, i4);
  })(e6, n6, t4) : i3(e6, n6);
}

// ../../node_modules/.pnpm/@lit+reactive-element@1.4.1/node_modules/@lit/reactive-element/node/decorators/state.js
function t3(t4) {
  return e4({ ...t4, state: true });
}

// ../../node_modules/.pnpm/@lit+reactive-element@1.4.1/node_modules/@lit/reactive-element/node/decorators/query-assigned-elements.js
var l4;
var n5 = null != (null === (l4 = globalThis.HTMLSlotElement) || void 0 === l4 ? void 0 : l4.prototype.assignedElements) ? (o6, l5) => o6.assignedElements(l5) : (o6, l5) => o6.assignedNodes(l5).filter((o7) => o7.nodeType === Node.ELEMENT_NODE);

export {
  __decorateClass,
  i,
  y,
  T,
  b,
  z,
  s4 as s,
  e4 as e,
  t3 as t
};
